Moving repo to fuel-telemetry (#1)
FuelLabs/fuel-telemetry: README.md

Reviewer: could be useful to have a docker command (or nix-shell command?) - to view/query the telemetry locally collected.
This would also help with debugging, testing and validating local collected traces - before sending them over the wire.

Author: I mentioned in last week's meeting that the InfluxDB URL endpoint can be set via an environment variable (i.e can be set to a locally running InfluxDB container), but that should also be documented here too, thanks!

Author: Cool, docs to run a local InfluxDB docker container has been added.

# Fuel Telemetry

`fuel-telemetry` is a `tracing` `Layer` used to implement telemetry within our
apps and libraries.

Steps to get `fuel-telemetry` going:

- Create a `TelemetryLayer` `tracing` `Layer`
- Add the `TelemetryLayer` to a `tracing` `Subscriber`
- Enable the `telemetry` field on `Span`s you want sent to InfluxDB
- Use the regular `tracing` macros (e.g `info!()`, `warn!()` etc) to record `Events` within `Span`s
- Create a `FileWatcher` so that telemetry files will be sent to InfluxDB

Most of these steps are hidden away behind convenience functions within `TelemetryLayer`.

## Using `fuel-telemetry`

### Within Applications with No Existing Tracing (Enabled by Default)

If you're writing an app where:

- `fuel-telemetry` will be your only `tracing` `Layer` and `Subscriber`
- and **you want all** `tracing` `Event`s to be sent to InfluxDB by default

then the quickest way to get going is:

> 
>     use fuel_telemetry::prelude::*;
> 
>     fn main() {
>         telemetry_init().unwrap();
> 
>         info!("This event will be sent to InfluxDB");
>     }

Reviewer: wrap in code block:
```suggestion
```rust
    use fuel_telemetry::prelude::*;

    fn main() {
        telemetry_init().unwrap();

        info!("This event will be sent to InfluxDB");
    }
```
```
^ closing "```" (not shown in suggestion above)

Reviewer: needs to be applied to all code blocks in readme.

        info!("This event will only be sent to stdout as telemetry=false by default");
    }

### Within Libraries

**Warning: libraries should not set the global default subscriber as it will
clobber any set by the application.**

Libraries should not need to interact with `TelemetryLayer` directly. Instead,
use the regular `tracing` crate macros (e.g `info!()`, `warn!()` etc) to record
`Event`s.

**Note:** `telemetry` is `disabled` by default. See below on how to enable telemetry
on `Event`s.

### Detailed Examples

For detailed examples, see `examples/*.rs`.

## Enabling Telemetry

Each `tracing` `Event` is either `enabled` or `disabled` based on the current
`Span`'s setting (if one exists). By default, the root `Span` is disabled with:

    telemetry=false

There are 2 ways to set the `telemetry` field:

1. On a `Span` itself e.g `span!(Level::ERROR, "span_name", telemetry=true)`
(see `examples/spans.rs`)
1. Via function attributes e.g `#[tracing::instrument(fields(telemetry=false))]`
(see `examples/attributes.rs`)

Given that `tracing` `Span`s are hierarchical, if a function does not have a
`tracing` attribute or does have one but the `telemetry` field is not set, we

> fall back by climing the heirarchy until we explicitly find one set. In the end

Reviewer 2: ```suggestion
fall back by climbing the hierarchy until we explicitly find one set. In the end
```

Author: Gah! Thanks for that

if no `Span` is found, we default to `telemetry=false`.

## Warning!

Be careful when using function attributes to enable telemetry as every function
call invocation, along with its parameter names and values, will be a recorded
within separate `Event`s and subsequently sent to InfluxDB...

* If used within a hot path, lots of data will be generated and sent over the
wire. Times this by the amount of installs across the entire userbase, and
you'll see that it could be an overwhelming amount of data and it's eventual
cost.

* PII (Personally Identifiable Information) and other sensitive information
  (think private keys and passwords) could easily make its way to InfluxDB by
  accident. If your functions have any PII (including Public Keys or even hashes
  of unique data), use the `skip()` function attribute function i.e:

  `#[tracing::instrument(skip(secret_key, password))]`

## Debugging Telemetry

### Inspecting Telemetry Files

> Telemetry files on disk are stored Base64 encoded. To peak into them:

Reviewer 2: ```suggestion
Telemetry files on disk are stored Base64 encoded. To peek into them:
```

## Using `fuel-telemetry`

### Within Applications with No Existing Tracing (Enabled by Default)

If you're writing an app where:

- `fuel-telemetry` will be your only `tracing` `Layer` and `Subscriber`
- and **you want all** `tracing` `Event`s to be sent to InfluxDB by default

then the quickest way to get going is:

```rust
use fuel_telemetry::prelude::*;

fn main() {
    telemetry_init().unwrap();

    info!("This event will be sent to InfluxDB");
}
```

### Within Applications with No Existing Tracing (Disabled by Default)

If you're writing an app where:
- `fuel-telemetry` will be your only `tracing` `Layer` and `Subscriber`
- and **you do NOT want all** `tracing` `Event`s to be sent to InfluxDB by default

then the quickest way to get going is:

```rust
use fuel_telemetry::prelude::*;

fn main() {
    let _guard = TelemetryLayer::new_global_default_with_watchers()?;
    info!("This event will not be sent in InfluxDB");

    // ... <code containing tracing events with telemetry disabled>
    // ...
    // ...

    // Time to enable telemetry from now on by setting telemetry=true on a `Span`
    let enabled_span = span!(Level::INFO, "root", telemetry=true);

>     let _span_guard = enabled_span.enter();

Reviewer: can we reduce this to 1 line?

Reviewer 3: https://docs.rs/tracing/latest/tracing/struct.Span.html#method.entered can be used to ge this in single line

```
let _span_guard = span!(Level::INFO, "root", telemetry=true).entered();
```

Author: I'm going to be adding convenience macros for each event type, so it would also make sense to make one here too.

Author: A `span_telemetry!()` macro along with `<level>_telemetry!()` macros have been added.

    info!("This event will be sent to InfluxDB");
}
```

### Within Applications with Existing Tracing

If you're writing an app where `fuel-telemetry` will need to work along side
other `tracing` `Layer`s or `Subscriber`s, you will need to create a
`TelemetryLayer` by hand and then add it to your existing `Subscriber` e.g:

```rust
use fuel_telemetry::prelude::*;
use tracing_subscriber::prelude::*;

fn main() {
    // Create a `Telemetry` `Layer` where events will appear in InfluxDB
    let (telemetry_layer, _guard) = TelemetryLayer::new_with_watchers().unwrap();

    // Create a stdout `Layer` where events will appear on `stdout`
    let stdout_layer = tracing_subscriber::fmt::layer().with_writer(std::io::stdout);

    // Create a `Subscriber` and combine these two `Layers`
    let subscriber = tracing_subscriber::registry()
        .with(telemetry_layer)
        .with(stdout_layer);

    // Set our subscriber as the default global subscriber

>     tracing::subscriber::set_global_default(subscriber).unwrap();

Reviewer 4: https://docs.rs/tracing/latest/tracing/index.html#in-executables

> Warning: In general, libraries should not call set_global_default()! Doing so will cause conflicts when executables that depend on the library try to set the default later.

Reviewer 3: Maybe we can wrap this one, and utilize type system to make downstream aware of this fact and make it a hard compile error if they try to use it in a library.

```rust
pub struct Binary;
pub struct Library;

pub trait CanSetGlobal {}
impl CanSetGlobal for Binary {}

pub fn set_global_subscriber<T>(
    _crate_type: T,
    subscriber: impl tracing::Subscriber + Send + Sync + 'static,
) where 
    T: CanSetGlobal,
{
    tracing::subscriber::set_global_default(subscriber).unwrap();
}
```

The downside of this is unfortunately there is no way to detect crate type in compile time.

https://github.com/rust-lang/rust/issues/20267

So this would be used like:

```rust
// In binary - compiles:
set_global_subscriber(Binary, my_subscriber);

// In library - compile error because `Library` doesn't implement `CanSetGlobal`:
set_global_subscriber(Library, my_subscriber);
```

Author: Interesting! I like the idea of using the type system to enforce rules :)

While there may not be a way to **detect** library vs binary at compile time, there is a way to influence the behaviour of the compiler between the two - by using `env!("CARGO_BIN_NAME")` - this will run fine for an end target, but will fail to compile for a library. I'll have a play with it.

Reviewer 3: Yeah i also looked into if it is possible to detect if it is binary or library before starting to build. There seemed to have no way. The `CARGO_BIN_NAME` sounds interesting! But i think it will be set after type checking. So it might not be useful for us in this scenario. That is why i fallback into the suggestion of using zero-sized structs to mark the crates using it. It could be possible to use build.rs and set a feature if that env is set but that would mean we need a build script only for this. Maybe it is not even possible with build.rs at all too in any case i think having the check through explicit types still utilizing the type system is a nice solution

Author: Constructor macros have now been added:
- `new!()`
- `new_with_watchers!()`
- `new_with_watchers_and_init!()`

Although only the last one needs to be `crate-type` gated, the other macros are still necessary as `env!("CARGO_PKG_NAME")` and `_VERSION` must be called from within the `fuel-telemetry`'s consumer's context in order to get the cargo target's end values... otherwise `env!("CARGO_PKG_NAME")` will forever return `fuel-telemetry` as it is in fact being called from within `fuel-telemetry`'s own calling context.

`tracing` attribute, or it does have one but the `telemetry` field is not set, we
fall back by climbing the hierarchy until we explicitly find one set. In the end
if no `Span` is found, we default to `telemetry=false`.

## Warning!

Be careful when using function attributes to enable telemetry as every function
call invocation (along with its parameter names and values) will be a recorded
as separate `Event`s and subsequently sent to InfluxDB...

* If used within a hot path, lots of data will be generated and sent over the
wire. Times this by the amount of installs across the entire userbase, and
you'll see that it could be an overwhelming amount of data and it's eventual
cost.

* PII (Personally Identifiable Information) and other sensitive information
  (think private keys and passwords) could make its way to InfluxDB by accident.
  If your functions have any PII (including Public Keys or even hashes of unique
  data etc), use the `skip()` function attribute i.e:

  `#[tracing::instrument(skip(secret_key, password))]`

## Debugging Telemetry

### Inspecting Telemetry Files

Telemetry files on disk are stored Base64 encoded. To peek into them:

>     cat ~/.fuelup/tmp/*.telemetry.* | while read line; do echo "$line" | base64 -d; echo; done

Reviewer: nit: wrap in markdown script:
```suggestion
```sh
cat ~/.fuelup/tmp/*.telemetry.* | while read line; do echo "$line" | base64 -d; echo; done
```
```

Author: Done.

### Manually Aging Out Telemetry Files

As `FileWatcher` only submits files to InfluxDB that are over an hour old by
default, we can force files to age out early so they are instantly submitted
on next run:

    for f in ~/.fuelup/tmp/*.telemetry.*; do mv "$f" "$f.old"; done

>     touch -t 202501010101 ~/.fuelup/tmp/*

Reviewer: nit: also wrap in markdown `sh` here too

Author: Done.

Most of these steps are hidden away behind convenience macros within `TelemetryLayer`.

## Using `fuel-telemetry`

### Convenience Macros

Within both applications and libraries, generate telemetry `Event`s using the
following convenience macros:

- `info_telemetry!()`
- `warn_telemetry!()`
- `error_telemetry!()`
- `debug_telemetry!()`
- `trace_telemetry!()`

When you want to record `tracing` `Event`s but don't want them to be sent to
InfluxDB, switch to using the regular non-`_telemetry` suffixed `tracing`
macros:

- `info!()`
- `warn!()`
- `error!()`
- `debug!()`
- `trace!()`

```rust
use fuel_telemetry::prelude::*;

info!("This event will be recorded");
info_telemetry!("This event will be recorded and sent to InfluxDB");
```

For detailed examples, see `examples/*.rs`.

### Applications without Existing Tracing

If you're writing an app where `fuel-telemetry` will be your only `tracing`
`Layer` and `Subscriber`, the quickest way to get going is:

```rust
use fuel_telemetry::prelude::*;

fn main() {
    let _guard = fuel_telemetry::new_with_watchers_and_init!().unwrap();

    info_telemetry!("This event will be sent to InfluxDB");
}
```

> ### Applications with Existing Tracing

Reviewer 2: Perhaps is would also be good to have a heading 

### Applications using `forc-tracing`
tiny hello world code ...

Author: Sounds good. Done in f7ae5058fd.

Reviewer 3: Seems to be addressed, closing the conversation.
